{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "956b7a64",
   "metadata": {},
   "source": [
    "# Classic predictions analysis tools\n",
    "\n",
    "This notebook contains some useful tools you can apply to check wether your predicted designs make sense, if they still contain your specified fixed_sequences, and/or how different they are from the original/WT proteins.\n",
    "\n",
    "**N.B.** - to use most of these functions, you will require to istall biopython - the first chunk of the notebook is there to make sure you have all dependencies installed.\n",
    "\n",
    "All these pipelines are written in **python** language.\n",
    "\n",
    "Should you encounter issues with the code or debugging, you can (1) google it; (2) ask your favourite AI friend, and, as a last resort, (3) ask one of the coaches to help you.\n",
    "\n",
    "\n",
    "Have a joyful analysis!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5b324f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ==========================================================\n",
    "# Importing needed libraries and dependencies\n",
    "# ==========================================================\n",
    "\n",
    "import os\n",
    "import numpy as np\n",
    "import glob\n",
    "import biopython\n",
    "\n",
    "# In case you want to make plots \n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69fb3b3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ==========================================================\n",
    "# 1. Extract reside types from input PDB file\n",
    "# ==========================================================\n",
    "\n",
    "def extract_residues_from_PDB(pdb_file):\n",
    "    \"\"\"Returns all non-consecutive duplicate residues (filters atom-level duplicates per residue, \n",
    "    keeps one per res), not \"unique\" residues. Returns full table: [(aa1_letter, chain, pos, bfactor), ...\"\"\"\n",
    "\n",
    "    # A dictionary mapping three-letter amino acid codes to one-letter codes\n",
    "    three_to_one = {\n",
    "        'ALA': 'A', 'CYS': 'C', 'ASP': 'D', 'GLU': 'E', 'PHE': 'F',\n",
    "        'GLY': 'G', 'HIS': 'H', 'ILE': 'I', 'LYS': 'K', 'LEU': 'L',\n",
    "        'MET': 'M', 'ASN': 'N', 'PRO': 'P', 'GLN': 'Q', 'ARG': 'R',\n",
    "        'SER': 'S', 'THR': 'T', 'VAL': 'V', 'TRP': 'W', 'TYR': 'Y'\n",
    "    }\n",
    "\n",
    "    # Initialize an empty list to store the amino acid positions\n",
    "    amino_acids = []\n",
    "\n",
    "    # Read the PDB file and extract the amino acid positions\n",
    "    with open(pdb_file, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('ATOM'):\n",
    "                amino_acid = line[17:20].strip()\n",
    "                strand_letter = line[21]\n",
    "                if amino_acid in three_to_one:\n",
    "                    amino_acid = three_to_one[amino_acid]\n",
    "                    position = int(line[22:26])\n",
    "                    confidence = int(float(line[60:66]))\n",
    "                    amino_acids.append((amino_acid, strand_letter, position, confidence))\n",
    "\n",
    "    # Remove consecutive duplicate entries from the amino acids list\n",
    "    residue_table = [] # filtered amino_acids\n",
    "    for i, entry in enumerate(amino_acids):\n",
    "        if i == 0 or entry[0:3] != amino_acids[i - 1][0:3]: # this caused a huge bug. In experimental PDB structures, the \"confidence\" varies from atom to atom\n",
    "            residue_table.append(entry)\n",
    "            \n",
    "    # Print the table of unique amino acid codes, strand letter, and positions\n",
    "    return residue_table\n",
    "\n",
    "\n",
    "\n",
    "def extract_contig_from_residue_table(residue_table,contigs):\n",
    "    \"\"\" After generating residue_table with extract_residues_from_PDB, this function takes \n",
    "    the contigs string as input and extracts the amino acids that are part of the contigs, \n",
    "    as well as the amino acids that are not part of the contigs. \n",
    "    It also returns a list of contigs as strings, and reconstructs contig sequences by \n",
    "    concatenating matching residues.\"\"\"\n",
    "\n",
    "    contigs_as_list_of_strings = []\n",
    "\n",
    "    # Process the entries in the input string\n",
    "    in_contig_amino_acids = []\n",
    "    for entry in contigs.split('/'):\n",
    "        if entry[0].isalpha():\n",
    "            letter = entry[0]\n",
    "            start, end = map(int, entry[1:].split('-'))\n",
    "            contig_string = \"\"\n",
    "            for amino_acid, strand_letter, position, confidence in residue_table:\n",
    "                if strand_letter == letter and start <= position <= end:\n",
    "                    in_contig_amino_acids.append((amino_acid, strand_letter, position, confidence))\n",
    "                    contig_string = contig_string + amino_acid\n",
    "            contigs_as_list_of_strings.append(contig_string)\n",
    "                    \n",
    "    # Remove in_contig_amino_acids from residue_table\n",
    "    variable_amino_acids = [row for row in residue_table if row not in in_contig_amino_acids]\n",
    "                    \n",
    "\n",
    "    return in_contig_amino_acids, variable_amino_acids, contigs_as_list_of_strings\n",
    "    \n",
    "    \n",
    "\n",
    "def index_contigs_in_generated_sequence(residue_table, contigs_as_list_of_strings):\n",
    "    \"\"\" This funtion takes the residue_table and the list of contigs as strings, and finds \n",
    "    the positions of the contigs in the amino acid sequence generated from the residue_table. \n",
    "    It Returns (matching_pos, non_matching_pos, matching_resIDs, confidences)—four lists of \n",
    "    indices of the amino acids that are part of the contigs, the amino acids that are not part \n",
    "    of the contigs, the residue IDs of the amino acids that are part of the contigs, and the \n",
    "    confidence scores of the amino acids that are part of the contigs.\"\"\"\n",
    "\n",
    "    # Extract the amino acid sequence from residue_table\n",
    "    sequence = \"\".join(row[0] for row in residue_table)\n",
    "\n",
    "    # Find positions and matching characters of consecutive strings in the sequence\n",
    "    positions = []\n",
    "    all_matching_positions = []\n",
    "    all_matching_positions_resIDs = []\n",
    "    confidences_of_residues = [] # snuck the list of confidences in here because would otherwise have to write exact same code all over again\n",
    "    for string in contigs_as_list_of_strings:\n",
    "        start = 0\n",
    "        while True:\n",
    "            position = sequence.find(string, start)\n",
    "            if position == -1:\n",
    "                break\n",
    "            positions.append(position + 1)  # Add 1 to match position in protein\n",
    "            matching_positions = range(position + 1, position + len(string) + 1)\n",
    "            all_matching_positions.extend(matching_positions)\n",
    "            for i_position in range(position,position+len(string)):\n",
    "                all_matching_positions_resIDs.append(residue_table[i_position][2])\n",
    "                confidences_of_residues.append(residue_table[i_position][3])\n",
    "            start = position + 1\n",
    "\n",
    "    # Find indices of characters not in all_matching_positions\n",
    "    not_matching_indices = [i + 1 for i, char in enumerate(sequence) if i + 1 not in all_matching_positions]\n",
    "\n",
    "    return all_matching_positions, not_matching_indices, all_matching_positions_resIDs, confidences_of_residues\n",
    "\n",
    "\n",
    "\n",
    "# residue_table = extract_residues_from_PDB(pdb_file)\n",
    "\n",
    "# in_contig_amino_acids, variable_amino_acids, contigs_as_list_of_strings = extract_contig_from_residue_table(residue_table,contigs)\n",
    "\n",
    "# all_matching_positions, not_matching_indices = index_contigs_in_generated_sequence(residue_table, contigs_as_list_of_strings)\n",
    "\n",
    "\n",
    "\n",
    "def parse_pdb_file(file_path):\n",
    "    \"\"\"Parse a PDB file and extract atom coordinates and residue IDs.\"\"\"\n",
    "\n",
    "    # Parse the PDB file and extract atom coordinates and residue IDs\n",
    "    atom_coords = []\n",
    "    residue_ids = []\n",
    "    with open(file_path, 'r') as pdb_file:\n",
    "        for line in pdb_file:\n",
    "            if line.startswith('ATOM'):\n",
    "                x = float(line[30:38])\n",
    "                y = float(line[38:46])\n",
    "                z = float(line[46:54])\n",
    "                residue_id = int(line[22:26]) # only number, no chain ID\n",
    "                atom_coords.append([x, y, z])\n",
    "                residue_ids.append(residue_id)\n",
    "    return np.array(atom_coords), residue_ids\n",
    "\n",
    "\n",
    "\n",
    "def compute_centroid(coords):\n",
    "    \"\"\"Computes the centroid of a set of coordinates.\"\"\"\n",
    "\n",
    "    # Compute the centroid of a set of coordinates\n",
    "    return np.mean(coords, axis=0)\n",
    "\n",
    "\n",
    "\n",
    "def compute_rmsd(coords1, coords2):\n",
    "    \"\"\" Computes the Root Mean Square Deviation (RMSD) between two sets of coordinates.\"\"\"\n",
    "\n",
    "    # Compute the RMSD between two sets of coordinates\n",
    "    diff = coords1 - coords2\n",
    "    rmsd = np.sqrt(np.mean(np.sum(diff**2, axis=1)))\n",
    "    return rmsd\n",
    "\n",
    "\n",
    "    \n",
    "def sort_atoms_by_type(pdb_file):\n",
    "    \"\"\" Sorts the atoms in a PDB file by atom type within each residue and removes hydrogen atoms.\"\"\"\n",
    "\n",
    "    # Open the PDB file\n",
    "    with open(pdb_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    # Filter and sort the atoms for each residue\n",
    "    residues = {}\n",
    "    current_residue = None\n",
    "    for line in lines:\n",
    "        if line.startswith('ATOM'):\n",
    "            residue_pos = int(line[22:26].strip())\n",
    "            atom_type = line[12:16].strip()\n",
    "\n",
    "            # Skip hydrogen atoms\n",
    "            if atom_type.startswith('H'):\n",
    "                continue\n",
    "\n",
    "            # Create a new residue entry\n",
    "            if residue_pos not in residues:\n",
    "                residues[residue_pos] = []\n",
    "\n",
    "            residues[residue_pos].append(line)\n",
    "\n",
    "    # Sort the atoms by atom type within each residue\n",
    "    sorted_lines = []\n",
    "    for residue_pos, atoms in sorted(residues.items()):\n",
    "        sorted_atoms = sorted(atoms, key=lambda x: x[12:16])\n",
    "        sorted_lines.extend(sorted_atoms)\n",
    "\n",
    "    # Save the sorted PDB file\n",
    "    output_file = pdb_file.split('.pdb')[0] + '_sorted.pdb'\n",
    "    with open(output_file, 'w') as file:\n",
    "        file.writelines(sorted_lines)\n",
    "\n",
    "    print(f\"Sorted PDB file saved as: {output_file}\")\n",
    "\n",
    "\n",
    "\n",
    "def rigid_alignment(ref_file, mov_file, ref_aa_names, mov_aa_names):\n",
    "    \"\"\"Performs rigid alignment of two PDB files based on specified amino acid \n",
    "    residues and computes the RMSD.\"\"\"\n",
    "\n",
    "    # Sort atoms in PDB files and delete hydrogens\n",
    "    sort_atoms_by_type(ref_file)\n",
    "    sort_atoms_by_type(mov_file)\n",
    "\n",
    "    # Parse the reference and mobile PDB files\n",
    "    ref_coords, ref_residues = parse_pdb_file(ref_file.split('.pdb')[0] + '_sorted.pdb')\n",
    "    mov_coords, mov_residues = parse_pdb_file(mov_file.split('.pdb')[0] + '_sorted.pdb')\n",
    "\n",
    "    # Select the specified amino acids for alignment\n",
    "    ref_indices = [i for i, res_id in enumerate(ref_residues) if res_id in ref_aa_names]\n",
    "    mov_indices = [i for i, res_id in enumerate(mov_residues) if res_id in mov_aa_names]\n",
    "\n",
    "    # Select the coordinates for the selected amino acids\n",
    "    ref_selected_coords = ref_coords[ref_indices]\n",
    "    mov_selected_coords = mov_coords[mov_indices]\n",
    "\n",
    "    # Compute the centroids of the selected sets\n",
    "    ref_centroid = compute_centroid(ref_selected_coords)\n",
    "    mov_centroid = compute_centroid(mov_selected_coords)\n",
    "\n",
    "    # Center the selected sets by subtracting the centroids\n",
    "    ref_centered_coords = ref_selected_coords - ref_centroid\n",
    "    mov_centered_coords = mov_selected_coords - mov_centroid\n",
    "     \n",
    "    # Compute the rotation matrix using singular value decomposition (SVD)\n",
    "    H = np.dot(ref_centered_coords.T, mov_centered_coords)\n",
    "    U, S, Vt = np.linalg.svd(H)\n",
    "    R = np.dot(Vt.T, U.T)\n",
    "\n",
    "    # Apply the rotation and translation to the mobile coordinates\n",
    "    rotated_mov_coords = np.dot(mov_coords - mov_centroid, R) + ref_centroid\n",
    "    \n",
    "    # Compute the RMSD between the specified sets of amino acids\n",
    "    rmsd = compute_rmsd(ref_selected_coords, rotated_mov_coords[mov_indices])\n",
    "\n",
    "    return rotated_mov_coords, rmsd\n",
    "\n",
    "\n",
    "\n",
    "def update_pdb_coordinates(input_file, output_file, new_coords):\n",
    "    \"\"\"Update the coordinates in a PDB file with new coordinates while preserving the \n",
    "    rest of the information. This function is useful in the context of rigid alignment, \n",
    "    where we want to update the coordinates of the mobile PDB file after alignment.\"\"\"\n",
    "\n",
    "    # Update the coordinates in the PDB file while preserving the rest of the information\n",
    "    with open(input_file, 'r') as input_pdb, open(output_file, 'w') as output_pdb:\n",
    "        for i, line in enumerate(input_pdb):\n",
    "            if line.startswith('ATOM'):\n",
    "                atom_line = line[:30] + f\"{new_coords[i, 0]:8.3f}{new_coords[i, 1]:8.3f}{new_coords[i, 2]:8.3f}\" + line[54:]\n",
    "                output_pdb.write(atom_line)\n",
    "            else:\n",
    "                output_pdb.write(line)\n",
    "\n",
    "\n",
    "\n",
    "def CompareTwoPDBs(contigs, ref_pdb_file, mov_pdb_file):\n",
    "    \"\"\" Compares two PDB files by performing rigid alignment based on specified amino acid \n",
    "    residues and computing the RMSD. It also returns the average, minimum, and maximum confidence \n",
    "    scores of the residues used for alignment.\"\"\"\n",
    "\n",
    "    res_table1 = extract_residues_from_PDB(ref_pdb_file)\n",
    "    contigs_as_list_of_strings = extract_contig_from_residue_table(res_table1,contigs)[2]\n",
    "    all_matching_positions_resIDs1 = index_contigs_in_generated_sequence(res_table1, contigs_as_list_of_strings)[2]\n",
    "\n",
    "    res_table2 = extract_residues_from_PDB(mov_pdb_file)\n",
    "    all_matching_positions_resIDs2 = index_contigs_in_generated_sequence(res_table2, contigs_as_list_of_strings)[2]\n",
    "    confidences_of_residues        = index_contigs_in_generated_sequence(res_table2, contigs_as_list_of_strings)[3]\n",
    "\n",
    "    # Perform rigid alignment and compute RMSD\n",
    "    rotated_coords, rmsd = rigid_alignment(ref_pdb_file, mov_pdb_file, all_matching_positions_resIDs1, all_matching_positions_resIDs2)\n",
    "\n",
    "    ## Update the coordinates in the mobile PDB file\n",
    "    update_pdb_coordinates(mov_pdb_file.split('.pdb')[0] + '_sorted.pdb', mov_pdb_file.split('.pdb')[0] + '_sorted-aligned.pdb', rotated_coords)\n",
    "\n",
    "    return rmsd, sum(confidences_of_residues)/len(confidences_of_residues), min(confidences_of_residues), max(confidences_of_residues)\n",
    "#    print(f\"RMSD: {rmsd:.3f}\")\n",
    "\n",
    "\n",
    "def parse_pdb_backbone_binders(pdb_file, chain_id='A'):\n",
    "    \"\"\" Parses a PDB file and extracts the coordinates of the backbone atoms (N, CA, C) for \n",
    "        a specified chain. This function is specifically designed for comparing the backbones \n",
    "        of two PDB files in the context of binders.\"\"\"\n",
    "\n",
    "    backbone_atoms = []\n",
    "    with open(pdb_file, \"r\") as file:\n",
    "        for line in file:\n",
    "            if line.startswith(\"ATOM\"):\n",
    "                atom_name = line[12:16].strip()\n",
    "                chain = line[21]\n",
    "                if atom_name in (\"N\", \"CA\", \"C\") and chain == chain_id:\n",
    "                    x = float(line[30:38])\n",
    "                    y = float(line[38:46])\n",
    "                    z = float(line[46:54])\n",
    "                    backbone_atoms.append((atom_name, (x, y, z)))\n",
    "    return backbone_atoms\n",
    "\n",
    "\n",
    "\n",
    "def align_pdb_backbones(reference_atoms, moving_atoms):\n",
    "    \"\"\" Aligns the backbone atoms of two PDB structures using the Kabsch algorithm.\"\"\"\n",
    "\n",
    "    if len(reference_atoms) != len(moving_atoms):\n",
    "        raise ValueError(\"The two structures have different numbers of backbone atoms.\")\n",
    "\n",
    "    # Convert atom coordinates to NumPy arrays\n",
    "    ref_coords = np.array([atom[1] for atom in reference_atoms])\n",
    "    mov_coords = np.array([atom[1] for atom in moving_atoms])\n",
    "\n",
    "    # Calculate the centroids for both structures\n",
    "    ref_centroid = np.mean(ref_coords, axis=0)\n",
    "    mov_centroid = np.mean(mov_coords, axis=0)\n",
    "\n",
    "    # Translate both structures to align centroids\n",
    "    ref_aligned = ref_coords - ref_centroid\n",
    "    mov_aligned = mov_coords - mov_centroid\n",
    "\n",
    "    # Calculate the rotation matrix\n",
    "    H = np.dot(ref_aligned.T, mov_aligned)\n",
    "    U, _, Vt = np.linalg.svd(H)\n",
    "    rotation_matrix = np.dot(Vt.T, U.T)\n",
    "\n",
    "    # Apply the rotation matrix to the aligned moving atoms\n",
    "    mov_rotated = np.dot(mov_aligned, rotation_matrix)\n",
    "\n",
    "    # Translate the rotated moving atoms back to the reference centroid\n",
    "    mov_final = mov_rotated + ref_centroid\n",
    "\n",
    "    # Convert back to list of tuples for output\n",
    "    aligned_atoms = [(atom[0], tuple(coord)) for atom, coord in zip(moving_atoms, mov_final)]\n",
    "\n",
    "    return aligned_atoms\n",
    "\n",
    "def calculate_rmsd_binders(reference_atoms, aligned_atoms):\n",
    "    \"\"\" Calculates the RMSD between the backbone atoms of two aligned PDB structures. \"\"\"\n",
    "    if len(reference_atoms) != len(aligned_atoms):\n",
    "        raise ValueError(\"The two structures have different numbers of backbone atoms.\")\n",
    "\n",
    "    squared_sum = sum(\n",
    "        np.sum((np.array(ref_coord) - np.array(aligned_coord)) ** 2) for (atom, ref_coord), (_, aligned_coord) in\n",
    "        zip(reference_atoms, aligned_atoms))\n",
    "    rmsd = np.sqrt(squared_sum / len(reference_atoms))\n",
    "\n",
    "    return rmsd\n",
    "\n",
    "def CompareTwoPDBs_binder(ref_pdb_file_path, mov_pdb_file_path):\n",
    "    \"\"\" Compares two PDB files by aligning their backbone atoms and calculating the RMSD.\"\"\"\n",
    "\n",
    "    reference_atoms = parse_pdb_backbone_binders(ref_pdb_file_path) \n",
    "    moving_atoms = parse_pdb_backbone_binders(mov_pdb_file_path)\n",
    "\n",
    "    if len(reference_atoms) != len(moving_atoms):\n",
    "        raise ValueError(\"The two structures have different numbers of backbone atoms.\")\n",
    "\n",
    "    # perform alignment only via the backbone\n",
    "    aligned_atoms = align_pdb_backbones(reference_atoms, moving_atoms)\n",
    "\n",
    "    # calculate RMSD of backbones\n",
    "    rmsd = calculate_rmsd_binders(reference_atoms, aligned_atoms)\n",
    "    print(f\"RMSD of backbones: {rmsd:.2f} Å\")\n",
    "\n",
    "    return rmsd"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
